core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1651,14 +1226,20 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
   private def triggerControllerMove(): Unit = {
     onControllerResignation()
     activeControllerId = -1
-    controllerContext.zkUtils.deletePath(ZkUtils.ControllerPath)
+    kafkaControllerZkUtils.deleteController()
+  }
+
+  def expire(): Unit = {
+    eventManager.clearAndPut(Expire)
```

Comment:
It's possible that the controller is still working on an event when expire() is called. In this case, it seems that we want to wait until the controller returns to idle state before we create a new ZK session. Otherwise, the controller may process an old event using the new session.